src/hackney_pool.erl

```diff
@@ -304,6 +307,16 @@ handle_cast({set_maxconn, MaxConn}, State) ->
 handle_cast({set_timeout, NewTimeout}, State) ->
   {noreply, State#state{timeout=NewTimeout}};
 
+handle_cast({checkout_cancel, Dest, Ref}, State) ->
+  {Queues, Removed} = del_from_queue(Dest, Ref, State#state.queues),
+  case Removed of
+    true ->
+      NbWaiters = State#state.nb_waiters - 1,
+      {noreply, State#state{queues=Queues, nb_waiters=NbWaiters}};
```

Comment:
why not dequeuing it there too?

Reply:
As we have non-zero `nb_waiters` only want `length(clients)==MaxConns` right, otherwise when one of those clients finished, `hackney_pool` would have dequeued already?

Reply:
releasing a waiter only happen when a connection is put back in the loop or when a request is canceled: forcefully closed via  hackney:close/1 or when a process actually requesting crash. I think we are in the second case there, so imo we should dequeue one of the waiter to actually tell him to create a new request imo.

Reply:
We will exceed `pool_size` though?

Reply:
@jimdigriz right. applying your change now. Will release sometimes today :)